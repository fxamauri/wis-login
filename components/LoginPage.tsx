import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { signInRequest } from '../actions';
import { AppState } from '../interfaces';
import * as S from './LoginPage.styles';
import Input from './Input';
import ButtonLogin from './ButtonLogin';
import LoginPageHeader from './LoginPageHeader';
import LoginPageForgotPassword from './LoginPageForgotPassword';

interface ErrorsState {
  email?: string;
  password?: string;
}

const LoginPage = () => {
  const dispatch = useDispatch();
  const loading = useSelector((state: AppState): boolean => state.loading);

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [errors, setErrors] = useState<ErrorsState>({});

  const validate = () => {
    const errors: ErrorsState = {};

    if (email.length === 0) {
      errors.email = 'O campo E-mail é obrigatório';
    } else if (!/\S+@\S+\.\S+/.test(email)) {
      errors.email = 'E-mail inválido';
    }

    if (password.length === 0) {
      errors.password = 'O campo senha é obrigatório';
    }

    setErrors(errors);
    return !(!!errors.email || !!errors.password);
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (validate()) {
      dispatch(signInRequest(email, password));
    }
  };

  return (
    <S.Container>
      <S.BackgroundWoman />
      <S.LoginContainer>
        <S.ToLeftMediumScreenSizes>
          <S.LoginForm onSubmit={handleSubmit} data-testid="login-form">
            <S.LoginContent>
              <LoginPageHeader
                title="Olá, seja bem vindo!"
                subTitle="Para acessar a plataforma, faça seu login."
              />
              <S.FormGroup>
                <Input
                  name="email"
                  label="E-mail"
                  type="text"
                  placeholder="user.name@mail.com"
                  isInvalid={!!errors.email}
                  invalidMessage={errors.email}
                  onChange={(e) => setEmail(e.target.value)}
                  value={email}
                />
              </S.FormGroup>
              <S.FormGroup>
                <Input
                  name="password"
                  label="Senha"
                  type="password"
                  placeholder="*******"
                  isInvalid={!!errors.password}
                  invalidMessage={errors.password}
                  onChange={(e) => setPassword(e.target.value)}
                  value={password}
                />
              </S.FormGroup>
              <S.ButtonLoginSection>
                <ButtonLogin title="Entrar" loading={loading} />
                <LoginPageForgotPassword />
              </S.ButtonLoginSection>
            </S.LoginContent>
          </S.LoginForm>
        </S.ToLeftMediumScreenSizes>
      </S.LoginContainer>
    </S.Container>
  );
};
export default LoginPage;
