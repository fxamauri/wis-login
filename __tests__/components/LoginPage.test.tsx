import { render, fireEvent } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';

import LoginPage from '../../components/LoginPage';

jest.mock('react-redux');

describe('LoginPage', () => {
  it('Should render loginPage', () => {
    const { container } = render(<LoginPage />);

    (useSelector as jest.Mock).mockImplementation((cb) =>
      cb({
        loading: false,
      }),
    );

    const dispatch = jest.fn();
    (useDispatch as jest.Mock).mockReturnValue(dispatch);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should be able to login', () => {
    const { getByPlaceholderText, getByTestId } = render(<LoginPage />);

    (useSelector as jest.Mock).mockImplementation((cb) =>
      cb({
        loading: true,
      }),
    );

    const dispatch = jest.fn();
    (useDispatch as jest.Mock).mockReturnValue(dispatch);

    fireEvent.change(getByPlaceholderText('user.name@mail.com'), {
      target: { value: 'email@email.com' },
    });

    fireEvent.change(getByPlaceholderText('*******'), {
      target: { value: 'astrongpassword' },
    });
    //
    fireEvent.submit(getByTestId('login-form'));
    //
    expect(dispatch).toHaveBeenCalled();
    // fireEvent.click(getByTestId('button-login'));
    // button-login
  });
});
