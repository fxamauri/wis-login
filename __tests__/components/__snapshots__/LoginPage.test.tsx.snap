// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Should render loginPage 1`] = `
<div
  class="LoginPagestyles__Container-sc-11ttscf-0 fvPDEz"
>
  <div
    class="LoginPagestyles__BackgroundWoman-sc-11ttscf-1 eURXJn"
  />
  <div
    class="LoginPagestyles__LoginContainer-sc-11ttscf-3 jAgnCy"
  >
    <div
      class="LoginPagestyles__ToLeftMediumScreenSizes-sc-11ttscf-2 calpJz"
    >
      <form
        class="LoginPagestyles__LoginForm-sc-11ttscf-4 eZhMGZ"
        data-testid="login-form"
      >
        <div
          class="LoginPagestyles__LoginContent-sc-11ttscf-5 hcCZaN"
        >
          <div
            class="LoginPageHeaderstyles__Container-sc-1kkzcr1-0 bxDIFd"
          >
            <h1
              class="LoginPageHeaderstyles__Title-sc-1kkzcr1-1 bWkUDc"
            >
              Olá, seja bem vindo!
            </h1>
            <p
              class="LoginPageHeaderstyles__SubTitle-sc-1kkzcr1-2 bFNnNa"
            >
              Para acessar a plataforma, faça seu login.
            </p>
          </div>
          <div
            class="LoginPagestyles__FormGroup-sc-11ttscf-6 jpMcGh"
          >
            <div
              class="Inputstyles__Container-sc-17xchd0-0 bOXAkU"
            >
              <label
                class="Inputstyles__Label-sc-17xchd0-1 iYXcqo"
                for="email"
              >
                E-mail
              </label>
              <div
                class="Inputstyles__InputContainer-sc-17xchd0-3 iUQGbd"
              >
                <input
                  class="Inputstyles__InputStyled-sc-17xchd0-2 ieXVSx"
                  data-testid="input"
                  name="email"
                  placeholder="user.name@mail.com"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="LoginPagestyles__FormGroup-sc-11ttscf-6 jpMcGh"
          >
            <div
              class="Inputstyles__Container-sc-17xchd0-0 bOXAkU"
            >
              <label
                class="Inputstyles__Label-sc-17xchd0-1 iYXcqo"
                for="password"
              >
                Senha
              </label>
              <div
                class="Inputstyles__InputContainer-sc-17xchd0-3 iUQGbd"
              >
                <input
                  class="Inputstyles__InputStyled-sc-17xchd0-2 ieXVSx"
                  data-testid="input"
                  name="password"
                  placeholder="*******"
                  type="password"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="LoginPagestyles__ButtonLoginSection-sc-11ttscf-7 ytbDH"
          >
            <button
              class="ButtonLoginstyles__ButtonLogin-d0aenn-0 kJBLfk"
            >
              Entrar
            </button>
            <span
              class="LoginPageForgotPasswordstyles__Container-sc-1ecj8hf-0 iNNOJL"
            >
              Esqueceu seu login ou senha? Clique
               
              <a
                href="/"
              >
                aqui
              </a>
            </span>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
